stan/math/prim/arr/functor/mpi_cluster.hpp

```diff
@@ -0,0 +1,251 @@
+   * the non-root processes we wait for broadcasts of mpi_command
+   * objects which are initiated on the root using the
+   * mpi_broadcast_command function below. Each recieved
+   * mpi_command is executed using the virtual run method.
+   */
+  void listen() {
+    mpi_cluster::cluster_listens_ = true;
+    if (rank_ == 0) {
+      return;
+    }
+    // TODO(wds15): Remove debugging print
+    std::cout << "Worker " << rank_ << " listening for commands..."
+              << std::endl;
+
+    try {
+      // lock on the workers the cluster as MPI commands must be
+      // initiated from the root and any attempt to do this on the
+      // workers must fail
+      std::unique_lock<std::mutex> worker_lock(in_use_);
+      while (1) {
+        boost::shared_ptr<mpi_command> work;
+
+        boost::mpi::broadcast(world_, work, 0);
+
+        work->run();
+      }
+    } catch (const mpi_stop_listen& e) {
+      // TODO(wds15): Remove debugging print
+      std::cout << "Wrapping up MPI on worker " << rank_ << " ..." << std::endl;
+    }
+  }
+
+  /**
+   * Stops listening state of the cluster. This happens
+   * automatically upon destruction. It is achieved by sending the
+   * mpi_stop_worker command.
+   */
+  void stop_listen() {
+    if (rank_ == 0 && mpi_cluster::cluster_listens_) {
+      mpi_broadcast_command<mpi_stop_worker>();
+    }
+    mpi_cluster::cluster_listens_ = false;
+  }
+
+  /**
+   * Returns the current listening state of the cluster.
+   */
+  static bool is_listening() { return cluster_listens_; }
+
+ private:
+  static bool cluster_listens_;
+};
+
+bool mpi_cluster::cluster_listens_ = false;
+std::mutex mpi_cluster::in_use_;
+
+/**
+ * Broadcasts a command instance to the listening cluster. This
+ * function must be called on the root whenever the cluster is in
+ * listening mode and errs otherwise.
+ *
+ * @param command shared pointer to an instance of a command class
+ * derived from mpi_command
+ * @return A unique_lock instance locking the mpi_cluster
+ */
+inline std::unique_lock<std::mutex> mpi_broadcast_command(
+    boost::shared_ptr<mpi_command> command) {
+  boost::mpi::communicator world;
+
+  if (world.rank() != 0)
+    throw std::runtime_error("only root may broadcast commands.");
+
+  if (!mpi_cluster::is_listening())
+    throw std::runtime_error("cluster is not listening to commands.");
+
+  std::unique_lock<std::mutex> cluster_lock(mpi_cluster::in_use_,
+                                            std::try_to_lock);
+
+  if (!cluster_lock.owns_lock())
+    throw mpi_is_in_use();
+
+  boost::mpi::broadcast(world, command, 0);
+
+  return cluster_lock;
+}
+
+/**
+ * Broadcasts default constructible commands to the cluster.
+ *
+ * @tparam T default constructible command class derived from
+ * mpi_command
+ * @return A unique_lock instance locking the mpi_cluster
+ */
+template <typename T>
+std::unique_lock<std::mutex> mpi_broadcast_command() {
+  boost::shared_ptr<mpi_command> command(new T);
```

Comment:
Where is the new `T` being freed?  Or is the intention that it's persistent forever?  It'd be nice ot have some doc here either way.

Reply:
The shared_ptr manages the lifetime of T. Whenever no more shared_ptr instance is pointing to T, then the last shared_ptr instance going out of scope will delete T. This is standard usage of shared_ptr's to my knowledge. I will add a comment.